src/main/java/bio/terra/stairway/Stairway.java

```diff
@@ -335,6 +562,38 @@ FlightDao getFlightDao() {
         return flightDao;
     }
 
+    // Exposed for work queue listener
+    ThreadPoolExecutor getThreadPool() {
+        return threadPool;
+    }
+
+    void exitFlight(FlightContext context)
+            throws DatabaseOperationException, FlightException, StairwayExecutionException, InterruptedException {
+        // save the flight state in the database
+        flightDao.exit(context);
+
+        if (context.getFlightStatus() == FlightStatus.READY && workQueueEnabled) {
+            queueFlight(context);
+        }
+    }
+
+    private void queueFlight(FlightContext context)
+            throws DatabaseOperationException, StairwayExecutionException, InterruptedException {
+        // If the flight state is READY, then we put the flight on the queue and mark it queued in the
+        // database. We cannot go directly from RUNNING to QUEUED. Suppose we put the flight in
+        // the queue before it was marked QUEUED. Another Stairway instance would see it was
+        // RUNNING and decide it shouldn't run it. Suppose we put the flight in the queue after
+        // we marked it QUEUED. Then if we fail before we put it on the queue, no one knows it should
+        // be queued. To close that window, we use the READY state to decouple the flight from RUNNING.
+        // Then we queue, then we mark as queued. Another Stairway looking for orphans, will find the READY
+        // and queue it. Putting a flight on the queue twice is not a problem. Stairway instances race to
```

Comment:
To make sure I'm understanding this right: putting a flight on the queue twice (or more) isn't a problem because one of the first steps to accepting a flight from the work queue (before doing any work) is marking ownership in the shared DB, and at-most one instance can do that because DBs are good at transaction management. Is that right?

Reply:
Exactly right.